src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp

```diff
@@ -0,0 +1,224 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "pch.h"
+#include "CppUnitTest.h"
+
+#include "ModernApp.h"
+#include "TestUtils.h"
+
+#include "UiaOperationAbstraction.h"
+
+using namespace UiaOperationAbstraction;
+
+namespace UiaOperationAbstractionTests
+{
+    // Detect whether a type can be returned from a remote operation. A user
+    // passes a single type parameter T in to CanBeReturned, and the type's
+    // underlying boolean value will represent whether the type can be returned
+    // or not.
+    //
+    // This works using a behavior called SFINAE. We define two overloads for
+    // CanBeReturned: one that inherits from true_type and one that inherits
+    // from false_type. When C++ is deciding which overload to choose, it
+    // starts with the true_type one, since it has the most specific signature.
+    // If the overload isn't ill-formed (i.e. the expression in decltype
+    // successfully compiles), the compiler will select that overload.
+    // Otherwise, it will move on to the false_type overload, which is never
+    // ill-formed. This means that the boolean value of the type is known at
+    // compile-time, and can be compared against an expected value.
+    //
+    // We use FromRemoteResult to test for whether the type can be returned
+    // because while calling BindResult on an un-returnable type would also
+    // fail to compile, the compiler can't detect compilation failures within
+    // that function. It can only reason about the flat expression in decltype.
+    template<class, class = std::void_t<>>
+    struct CanBeReturned : std::false_type{};
+
+    template<class T>
+    struct CanBeReturned<T,
+            std::void_t<decltype(std::declval<T>().FromRemoteResult(std::declval<winrt::Windows::Foundation::IInspectable>()))>
+        > : std::true_type{};
+
+    TEST_CLASS(UiaOperationAbstractionTests)
+    {
+    public:
+        // Asserts that you can get the name of a UiaElement.
+        void ElementGetNameTest(const bool useRemoteOperations)
+        {
+            ModernApp app(L"Microsoft.WindowsCalculator_8wekyb3d8bbwe!App");
+            app.Activate();
+            auto calc = WaitForElementFocus(L"Display is 0");
+
+            winrt::com_ptr<IUIAutomation> automation;
+            THROW_IF_FAILED(::CoCreateInstance(__uuidof(CUIAutomation8), nullptr, CLSCTX_INPROC_SERVER, IID_PPV_ARGS(automation.put())));
+            UiaOperationAbstraction::Initialize(useRemoteOperations, automation.get());
+            auto cleanup = wil::scope_exit([&]()
```

Comment:
Doesn't need any concrete capture since it just calls the global cleanup function.

Reply:
This is fine as it is. It's always safe to use a blanket by-ref capture with `scope_exit` and there's no need to think about what to capture all the time; just consistently use by ref and it's always correct.

---
In reply to: [460283878](https://github.com/microsoft/Microsoft-UI-UIAutomation/pull/10#discussion_r460283878) [](ancestors = 460283878)

Reply:
Discussed offline: Let's not do it. :)

Reply:
When moving this to a function I got rid of the capture, since when returning a `wil::scope_exit` from a function it definitely shouldn't reference local state. Thanks for the feedback :)